### PR TITLE
clip element support for wdio

### PR DESCRIPTION
**Pull request by kb-kerem**
> Issue : IRIS-1033

## Description
Adds missing clip element support for wdio

## Types of Changes
- New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: visual-js/visual-wdio/src/SauceVisualService.ts [hunk 1]
```diff
@@ -406,6 +409,8 @@ export default class SauceVisualService implements Services.ServiceInstance {
       const jobId = (browser.capabilities as any)['jobUuid'] || sessionId;
       const result = await api.createSnapshotFromWebDriver({
         captureDom: options.captureDom ?? this.captureDom,
+        clipElement:
+          options.clipElement?.elementId ?? this.clipElement?.elementId,
         clipSelector: options.clipSelector ?? this.clipSelector,
```
Comment (FriggaHel) [inline visual-js/visual-wdio/src/SauceVisualService.ts:414]: Should we at the same time make the `clipSelector` use `clipElement` behind the scene?

@omacranger what do you think ?

Comment (omacranger) [reply visual-js/visual-wdio/src/SauceVisualService.ts:415]: Yeah I think that would be fine to do now. They would both throw the same exception if it couldn't be found / resolved on the backend so it wouldn't be a breaking behavior.

File: visual-js/visual-wdio/src/SauceVisualService.ts [hunk 2]
```diff
@@ -118,6 +119,7 @@ export type CheckOptions = {
    * A querySelector compatible selector of an element that we should crop the screenshot to.
    */
   clipSelector?: string;
+  clipElement?: WdioElement;
```
[the remark this answers is not in the document]
Comment (omacranger) [reply visual-js/visual-wdio/src/SauceVisualService.ts:122]: Eh, I'm not sure if these types are super helpful. Technically they make them more accurate but the types are then far less legible, and we'd be likely deprecating clipSelector at some point in the future anyway. Even a docblock comment like `clipElement takes priority over a clipSelector` would be easy enough for me.

File: visual-js/visual-wdio/src/SauceVisualService.ts [hunk 3]
```diff
@@ -404,9 +408,19 @@ export default class SauceVisualService implements Services.ServiceInstance {
 
       const sessionId = browser.sessionId;
       const jobId = (browser.capabilities as any)['jobUuid'] || sessionId;
+
+      let clipSelector = options.clipSelector ?? this.clipSelector;
+      if (clipSelector) {
+        const element = await browser.$(clipSelector);
+        clipSelector = element.elementId;
+      }
```
Comment (omacranger) [inline visual-js/visual-wdio/src/SauceVisualService.ts:416]: Perhaps this instead? IMO I'd try not to mutate variables and reassign them to a value that doesn't match the name  (`clipElement` in a `clipSelector`)

```
const clipSelector = options.clipSelector ?? this.clipSelector;
const clipElement = clipSelector ? await browser.$(clipSelector) : null;
```

File: visual-js/visual-wdio/src/SauceVisualService.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (kb-kerem) [reply visual-js/visual-wdio/src/SauceVisualService.ts:125]: I went for the comment way. @FriggaHel let me know if we still should enforce the types.

File: visual-js/visual-wdio/src/SauceVisualService.ts [hunk 4]
```diff
@@ -156,6 +158,7 @@ export default class SauceVisualService implements Services.ServiceInstance {
   diffingMethod: DiffingMethod | undefined;
   captureDom: boolean | undefined;
   clipSelector: string | undefined;
+  clipElement: WdioElement | undefined;
```
[the remark this answers is not in the document]
Comment (kb-kerem) [reply visual-js/visual-wdio/src/SauceVisualService.ts:164]: See previous comment

**Review by omacranger — APPROVED** [review]

File: visual-js/visual-wdio/src/SauceVisualService.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (FriggaHel) [reply visual-js/visual-wdio/src/SauceVisualService.ts:125]: IMO that's always better to keep impossible situations impossible. That seems too late anyway.

**Pull request merged**